src/guide/agent-functions/defining-agent-functions.rst

```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+
+.. _Pure Python Agent Functions:
+
+FLAME GPU Pure Python Agent Functions
+-------------------------------------
+
+Ane experimental feature allows the specification of *pure* Python agent functions. In this case agent functions can be expressed using a limited subset of Python. This is limited to syntax which can be transpiled to a C++ function at runtime for compilation. For example Python functionality like generator expressions, arrays, dictionaries, etc, which do not have an obvious C++ equivalent (excluding the in the SDL which is not supported in agent functions) is not permitted and will raise an exception. 
+
+Pure Python agent functions are required to have the ``@pyflamegpu.agent_function`` decorator and must specify two arguments, the message in message out variable, both of which must use a type annotation of a :ref:`supported message type<Communication Strategies>` prefixed with the Python FLAME GPU module name ``pyflamegpu.``.
+
+.. tabs::
+
+  .. code-tab:: py Python
+
+    #Define an agent function called agent_fn1
+    @pyflamegpu.agent_function
+    def outputdata(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageNone):
+        # Behaviour goes here
+        pass
+
+    ...
+    # Transpile the Pure Python agent function to equivalent C++ (and transpile errors with raise an exception at this stage)
+    agent_fn1_translated = pyflamegpu.codegen.translate(agent_fn1) 
+
+    # Attach pure Python function called agent_fn1 to an agent represented by the AgentDescription agent (the function will be compiled at runtime as C++)
+    agent.newRTCFunction("agent_fn1", agent_fn1_translated)
+    ...
+    
+Supported Function Calls
+""""""""""""""""""""""""
+
+The pure Python transpiler supports function calls that have a C++ equivalent in the will check function calls in the API. The API singleton is available as ``pyflamegpu`` which is equivalent to the ``FLAMEGPU`` object in C++. If non existent functions are called on API objects then this will raise an error during translation. Calling of correctly specified device functions is also supported as well as calls to a small number of python built in functions. E.g. ``abs()``, ``int()``, and ``float()``. Any calls to math library functions are directly translated to C++ equivalents. E.g. ``Math.sinf()`` will result in a call to ``sinf()`` in C++. The majority of python math functions directly translate to their C++ counterparts. 
+
+Supported Math Constants
+""""""""""""""""""""""""
+
+The following mathematics constants are supported;
+
+============= =================================
+Python        Description
+============= =================================
+``math.pi``   Translated of ``M_PI`` in C++
+``math.e``    Translated of ``M_E`` in C++
+``math.inf``  Translated of ``INFINITY`` in C++
+``math.nan``  Translated of ``NAN`` in C++
```

Comment:
How does Python handle NaN comparison?

NaN can only be tested with `isnan()` in C++.

Reply:
(Or comparing the variable with itself `a!=a` apparently evaluateS true if a is NaN.)

Reply:
```
>>> import math
>>> a = math.nan
>>> a != a
True
```
Applies to Python too, so nan without `isnan` seems pretty useless.

Reply:
You could use `isnan` in pythin by calling `isnan(x)` which would translate directly to the same call in C++ (but generate a warning) or `math.isnan` which would translate to `isnan(x)` (with no warning). 

Its not *pure* Python but you its probably worth leaving in as you can generate functional C++ using it.